### PR TITLE
Change invalid opcode to revert for input validation.

### LINKED ISSUE #2293: Encoding of preconditions for non-payable functions
According to the documentation (https://github.com/ethereum/solidity/blob/develop/docs/control-structures.rst), the compiler introduces an `invalid` instruction for the following two scenarios:
- "If your contract receives Ether via a public function without payable modifier (including the constructor and the fallback function)."
- "If your contract receives Ether via a public getter function."

Given that in such cases the callee can't really prevent callers from passing Ether, I am tempted to say that the compiler should instead introduce a `revert` instruction. In some sense, this is a kind of implicit parameter validation or precondition that is automatically inserted by the compiler.

@chriseth @pirapira @retotrinkler What do you think?

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #2293 and #2241

### CODE REVIEW

**Comment by axic** [comment]
There is still another case in `convertType` using the invalid opcode.

Also assembly, sourcelocation and bytecode tests are failing.

**Review by axic — CHANGES_REQUESTED** [review]
Need to change the other case in `convertType` to revert and update the docs.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -586,7 +586,7 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 			m_context << Instruction::CREATE;
 			// Check if zero (out of stack or not enough balance).
 			m_context << Instruction::DUP1 << Instruction::ISZERO;
-			m_context.appendConditionalInvalid();
+			m_context.appendConditionalRevert();
```
Comment (wuestholz) [inline libsolidity/codegen/ExpressionCompiler.cpp:590]: @chriseth I'm not completely sure about this case. Is the reasoning here that the caller has no way to prevent this failure and therefore requires that it holds from then on?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -650,7 +650,7 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 			{
 				// Check if zero (out of stack or not enough balance).
 				m_context << Instruction::ISZERO;
-				m_context.appendConditionalInvalid();
+				m_context.appendConditionalRevert();
```
Comment (wuestholz) [inline libsolidity/codegen/ExpressionCompiler.cpp:654]: @chriseth I'm also not completely sure about this case. Is the reasoning here similar to the one for contract creations?

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp:590]: This is part of contract creation (also the below cases, where transfer failed, code not present, call failed). I think the original agreement was these are invalid as they are not part of input validation.

Though one could argue that "code not present" is part of input validation.

Maybe we need more clarified rules.

**Comment by axic** [comment]
Also need to document in http://solidity.readthedocs.io/en/develop/control-structures.html#exceptions that it is thrown if a method is not found and no fallback is defined.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 3]
```diff
@@ -1694,7 +1692,7 @@ void ExpressionCompiler::appendExternalFunctionCall(
 	if (funKind == FunctionType::Kind::External || funKind == FunctionType::Kind::CallCode || funKind == FunctionType::Kind::DelegateCall)
 	{
 		m_context << Instruction::DUP1 << Instruction::EXTCODESIZE << Instruction::ISZERO;
-		m_context.appendConditionalInvalid();
+		m_context.appendConditionalRevert();
```
Comment (wuestholz) [inline libsolidity/codegen/ExpressionCompiler.cpp:1696]: @chriseth I'm also not completely sure about this case. Is the reasoning that the caller can't really check and prevent this scenario explicitly?

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:590]: The failure could be due to out-of-gas, which is an external parameter.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:654]: Yes, similar, but also this depends on other factors like timing and so on. Happy to discuss all these!

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1696]: Yes, the target contract could have selfdestructed, for example. If we use `Invalid` here, you cannot ever interact with contracts that might selfdestruct.

**Comment by chriseth** [comment]
Concerning the type conversion cases: One of them is integer to enum conversion which (for what I have seen in the code) will not be used for parsing input arguments as there, we only have `convertType(type, type, true)`.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (wuestholz) [reply libsolidity/codegen/ExpressionCompiler.cpp:590]: Ok, that's sort of what I assumed. Thanks!

Comment (wuestholz) [reply libsolidity/codegen/ExpressionCompiler.cpp:590]: I guess there are two types of inputs that might require validation. Obviously, there are method parameters that the caller provides. However, I would include return values of callee methods as well.

For instance, method A might call a method B and expect the return value to be non-zero (e.g., to prevent division-by-zero later in the code). Since method A cannot always control the return value of B it seems fine to revert in such cases.

In other words, it's not a bug in A if B returns 0. Does that make sense?

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 3 above)
Comment (wuestholz) [reply libsolidity/codegen/ExpressionCompiler.cpp:1696]: Ok, that makes sense. Thanks!

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 2 above)
Comment (wuestholz) [reply libsolidity/codegen/ExpressionCompiler.cpp:654]: I see. I assume that those factors are out of the control of the caller, right?

**Comment by chriseth** [comment]
Pending work:
 - [x] update documentation around exceptions
 - [x] clarify description of require and assert / revert and invalid opcode
 - [x] fix tests

**Comment by chriseth** [comment]
Finished from my side (as long as travis agrees).

File: docs/control-structures.rst [hunk 4]
```diff
@@ -361,55 +361,73 @@ As a result, the following code is legal, despite being poorly written::
...
+            // Since transfer throws an exception on failure and
+            // cannot call back here, there should be no way for us to
+            // still have half of the money.
+            assert(this.balance == balanceBeforeTransfer - msg.value / 2);
             return this.balance;
         }
     }
 
-Currently, Solidity automatically generates a runtime exception in the following situations:
+Currently, Solidity automatically generates an exception in the following situations. These exceptions
```
Comment (axic) [inline docs/control-structures.rst]: This line can be reworked (it suggests ``assert`` style is generated, while the other isn't)

File: docs/control-structures.rst [hunk 5]
```diff
@@ -361,55 +361,73 @@ As a result, the following code is legal, despite being poorly written::
...
+(instruction ``0xfe``) to throw an ``assert``-style exception. In both cases, this causes
+the EVM to revert all changes made to the state. The reason for reverting is that there is no safe way to continue execution, because an expected effect
 did not occur. Because we want to retain the atomicity of transactions, the safest thing to do is to revert all changes and make the whole transaction
-(or at least call) without effect.
-
-If contracts are written so that ``assert`` is only used to test internal conditions and ``require``
-is used in case of malformed input, a formal analysis tool that verifies that the invalid
-opcode can never be reached can be used to check for the absence of errors assuming valid inputs.
+(or at least call) without effect. Note that ``assert``-style exceptions consume all gas available to the call, while
+``revert``-style exceptions will not consume any gas starting from the metropolis release.
```
Comment (axic) [inline docs/control-structures.rst]: Metropolis with capital M

File: docs/control-structures.rst [hunk 6]
```diff
@@ -361,55 +361,73 @@ As a result, the following code is legal, despite being poorly written::
...
+and ``require`` should be used to check external conditions (invalid inputs or errors in external components).
+The idea behind that is that analysis tools can check your contract and try to come up with situations and
+series of function calls that will reach a failing assertion. If this is possible, this means there is a bug
+in your contract you should fix.
 
-There are some cases where exceptions are thrown automatically (see below). You can use the ``throw`` instruction to throw an exception manually. The effect of an exception is that the currently executing call is stopped and reverted (i.e. all changes to the state and balances are undone) and the exception is also "bubbled up" through Solidity function calls (exceptions are ``send`` and the low-level functions ``call``, ``delegatecall`` and ``callcode``, those return ``false`` in case of an exception).
+There are two other ways to trigger execptions: The ``revert`` function can be used to flag an error and
+revert the current call. In the future, it migt be possible, to also include details about the error
```
Comment (axic) [inline docs/control-structures.rst]: migt -> might

File: docs/units-and-global-variables.rst [hunk 7]
```diff
@@ -79,13 +79,23 @@ Block and Transaction Properties
     You can only access the hashes of the most recent 256 blocks, all other
     values will be zero.
 
-.. index:: assert, revert, keccak256, ripemd160, sha256, ecrecover, addmod, mulmod, cryptography, this, super, selfdestruct, balance, send
```
Comment (axic) [inline docs/units-and-global-variables.rst]: The index is missing for the crypto functions.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 8]
```diff
@@ -889,7 +887,7 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				// condition was not met, flag an error
 				m_context << Instruction::INVALID;
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Use `appendInvalid` here?

File: test/libsolidity/JSONCompiler.cpp [hunk 9]
```diff
@@ -75,26 +75,34 @@ BOOST_AUTO_TEST_CASE(basic_compilation)
 	BOOST_CHECK(contract["interface"].isString());
 	BOOST_CHECK(contract["interface"].asString() == "[]");
```
Comment (axic) [inline test/libsolidity/JSONCompiler.cpp]: Could use `BOOST_CHECK_EQUAL`.

File: docs/control-structures.rst (hunk 4 above)
Comment (chriseth) [reply docs/control-structures.rst]: I don't understand. The other states that a `require`-style exception is generated. Do you have an suggestion for rewording?

Comment (axic) [reply docs/control-structures.rst]: Pushed a commit.

**Comment by axic** [comment]
Needs rebase (also just in case any other merged changes would change tests).

**Comment by chriseth** [comment]
@axic can this be merged?

**Comment by axic** [comment]
@chriseth see comment above about rebase 😉

**Comment by chriseth** [comment]
@axic I think travis runs the tests on the merged commit and not on the pull request tip, but I might be wrong. Rebasing.

**Comment by chriseth** [comment]
Ah, it already was rebased.

**Comment by axic** [comment]
@chriseth tests are failing :(

**Review by axic — APPROVED** [review]

**Comment by chriseth** [comment]
Ok, this was hopefully the final fix.
